flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java

```diff
@@ -65,33 +72,46 @@ public void open(
 
-        this.checkpointedState =
-                context.getOperatorStateStore()
-                        .getListState(
-                                new ListStateDescriptor<>(
-                                        name + "-sequence-state", LongSerializer.INSTANCE));
-        this.valuesToEmit = new ArrayDeque<>();
-        if (context.isRestored()) {
-            // upon restoring
+        ListStateDescriptor<InternalState> stateDescriptor =
+                new ListStateDescriptor<>(
+                        name + "-sequence-state", TypeInformation.of(InternalState.class));
+        this.checkpointedState = context.getOperatorStateStore().getListState(stateDescriptor);
+        this.internalStates = Lists.newArrayList();
 
-            for (Long v : this.checkpointedState.get()) {
-                this.valuesToEmit.add(v);
-            }
+        totalNoOfElements = Math.abs(end - start + 1);
+        if (context.isRestored()) {
+            checkpointedState.get().forEach(state -> internalStates.add(state));
         } else {
             // the first time the job is executed
-            final int stepSize = runtimeContext.getNumberOfParallelSubtasks();
             final int taskIdx = runtimeContext.getIndexOfThisSubtask();
-            final long congruence = start + taskIdx;
+            final long stepSize = runtimeContext.getNumberOfParallelSubtasks();
+            internalStates.add(new InternalState(0, taskIdx, stepSize));
+        }
+    }
 
-            long totalNoOfElements = Math.abs(end - start + 1);
-            final int baseSize = safeDivide(totalNoOfElements, stepSize);
-            final int toCollect =
-                    (totalNoOfElements % stepSize > taskIdx) ? baseSize + 1 : baseSize;
+    private long toCollect(long baseSize, long stepSize, int taskIdx) {
+        return (totalNoOfElements % stepSize > taskIdx) ? baseSize + 1 : baseSize;
+    }
 
-            for (long collected = 0; collected < toCollect; collected++) {
-                this.valuesToEmit.add(collected * stepSize + congruence);
+    public Long nextValue() {
+        Iterator<InternalState> iterator = internalStates.iterator();
+        if (iterator.hasNext()) {
+            InternalState state = iterator.next();
+            long nextSequence = state.collected * state.stepSize + (start + state.taskId);
+            state.collected++;
+            // All sequence values are cleared from the stateList after they have been sent
+            if (state.collected
+                    >= toCollect(
+                            safeDivide(totalNoOfElements, state.stepSize),
+                            state.stepSize,
+                            state.taskId)) {
+                iterator.remove();
             }
+            return nextSequence;
         }
+
+        // Before calling this method, you should call hasNext to check
+        throw new IllegalStateException();
```

Comment:
```suggestion
        throw new IllegalStateException("SequenceGenerator.nextValue() was called with no remaining values.");
```
This code should be unreachable, so it's not a big deal -- but just in case, a helpful message is a good practice.

Reply:
You're quite right. I'll change it.